server/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java

```diff
@@ -651,32 +673,56 @@ public ClusterState execute(BatchExecutionContext<StartedShardUpdateTask> batchE
+                                startedShardEntry.shardId,
+                                matched,
+                                startedShardEntry
+                            );
                             tasksToBeApplied.add(taskContext);
                             shardRoutingsToBeApplied.add(matched);
                             seenShardRoutings.add(matched);
 
-                            // expand the timestamp range recorded in the index metadata if needed
-                            final Index index = entry.shardId.getIndex();
-                            IndexLongFieldRange currentTimestampMillisRange = updatedTimestampRanges.get(index);
+                            // expand the timestamp range(s) recorded in the index metadata if needed
+                            final Index index = startedShardEntry.shardId.getIndex();
+                            ClusterStateTimeRanges clusterStateTimeRanges = updatedTimestampRanges.get(index);
+                            IndexLongFieldRange currentTimestampMillisRange = clusterStateTimeRanges == null
+                                ? null
+                                : clusterStateTimeRanges.timestampRange();
+                            IndexLongFieldRange currentEventIngestedMillisRange = clusterStateTimeRanges == null
+                                ? null
+                                : clusterStateTimeRanges.eventIngestedRange();
+
                             final IndexMetadata indexMetadata = initialState.metadata().index(index);
                             if (currentTimestampMillisRange == null) {
                                 currentTimestampMillisRange = indexMetadata.getTimestampRange();
                             }
-                            final IndexLongFieldRange newTimestampMillisRange;
-                            newTimestampMillisRange = currentTimestampMillisRange.extendWithShardRange(
-                                entry.shardId.id(),
+                            if (currentEventIngestedMillisRange == null) {
+                                currentEventIngestedMillisRange = indexMetadata.getEventIngestedRange();
+                            }
+
+                            final IndexLongFieldRange newTimestampMillisRange = currentTimestampMillisRange.extendWithShardRange(
+                                startedShardEntry.shardId.id(),
                                 indexMetadata.getNumberOfShards(),
-                                entry.timestampRange
+                                startedShardEntry.timestampRange
                             );
-                            if (newTimestampMillisRange != currentTimestampMillisRange) {
-                                updatedTimestampRanges.put(index, newTimestampMillisRange);
+                            final IndexLongFieldRange newEventIngestedMillisRange = currentEventIngestedMillisRange.extendWithShardRange(
```

Comment:
I think we should only track this if `batchExecutionContext.initialState().getMinTransportVersion().onOrAfter(TransportVersions.EVENT_INGESTED_RANGE_IN_CLUSTER_STATE)`. Otherwise it should be `UNKNOWN`. If we don't do that, we will have different cluster states on different nodes because we can't send this data over the wire to older nodes.

Reply:
Changed in next push.